File: agentloop/events.go
package agentloop

import (
	"sync"
	"time"
)

// EventKind identifies the type of session event.
type EventKind string

const (
	EventSessionStart        EventKind = "session_start"
	EventSessionEnd          EventKind = "session_end"
	EventUserInput           EventKind = "user_input"
	EventAssistantTextStart  EventKind = "assistant_text_start"
	EventAssistantTextDelta  EventKind = "assistant_text_delta"
	EventAssistantTextEnd    EventKind = "assistant_text_end"
	EventToolCallStart       EventKind = "tool_call_start"
	EventToolCallOutputDelta EventKind = "tool_call_output_delta"
	EventToolCallEnd         EventKind = "tool_call_end"
	EventSteeringInjected    EventKind = "steering_injected"
	EventTurnLimit           EventKind = "turn_limit"
	EventLoopDetection       EventKind = "loop_detection"
	EventWarning             EventKind = "warning"
	EventError               EventKind = "error"
)

// SessionEvent is a typed event emitted by the agent loop.
type SessionEvent struct {
	Kind      EventKind              `json:"kind"`
	Timestamp time.Time              `json:"timestamp"`
	SessionID string                 `json:"session_id"`
	Data      map[string]interface{} `json:"data,omitempty"`
}

// EventEmitter delivers typed events to the host application via a channel.
type EventEmitter struct {
	sessionID string
	ch        chan SessionEvent
	closed    bool
	mu        sync.Mutex
}

// NewEventEmitter creates a new EventEmitter with a buffered channel.
func NewEventEmitter(sessionID string, bufferSize int) *EventEmitter {
	if bufferSize <= 0 {
		bufferSize = 256
	}
	return &EventEmitter{
		sessionID: sessionID,
		ch:        make(chan SessionEvent, bufferSize),
	}
}

// Emit sends an event to the channel. If the emitter is closed, the event
// is silently dropped.
func (e *EventEmitter) Emit(kind EventKind, data map[string]interface{}) {
	e.mu.Lock()
	defer e.mu.Unlock()
	if e.closed {
		return
	}
	event := SessionEvent{
		Kind:      kind,
		Timestamp: time.Now(),
		SessionID: e.sessionID,
		Data:      data,
	}
	select {
	case e.ch <- event:
	default:
		// Channel full; drop event to avoid blocking the agent loop.
	}
}

// Events returns the read-only event channel.
func (e *EventEmitter) Events() <-chan SessionEvent {
	return e.ch
}

// Close closes the event channel. Safe to call multiple times.
func (e *EventEmitter) Close() {
	e.mu.Lock()
	defer e.mu.Unlock()
	if !e.closed {
		e.closed = true
		close(e.ch)
	}
}

File: packages/attractor-agent/src/attractor_agent/events.py
"""Session event types and emitter."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Callable


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class EventKind(Enum):
    SESSION_START = "session_start"
    SESSION_END = "session_end"
    USER_INPUT = "user_input"
    ASSISTANT_TEXT_START = "assistant_text_start"
    ASSISTANT_TEXT_DELTA = "assistant_text_delta"
    ASSISTANT_TEXT_END = "assistant_text_end"
    TOOL_CALL_START = "tool_call_start"
    TOOL_CALL_OUTPUT_DELTA = "tool_call_output_delta"
    TOOL_CALL_END = "tool_call_end"
    STEERING_INJECTED = "steering_injected"
    TURN_LIMIT = "turn_limit"
    LOOP_DETECTION = "loop_detection"
    ERROR = "error"


@dataclass
class SessionEvent:
    kind: EventKind
    session_id: str
    data: dict[str, Any] = field(default_factory=dict)
    timestamp: datetime = field(default_factory=_utcnow)


class EventEmitter:
    def __init__(self):
        self._subscribers: list[Callable[[SessionEvent], None]] = []

    def subscribe(self, callback: Callable[[SessionEvent], None]) -> None:
        self._subscribers.append(callback)

    def emit(
        self, kind: EventKind, session_id: str, data: dict[str, Any] | None = None
    ) -> SessionEvent:
        event = SessionEvent(kind=kind, session_id=session_id, data=data or {})
        for subscriber in list(self._subscribers):
            subscriber(event)
        return event
